### PR TITLE
[JUJU-1887] Adds kill delay duration for service shutdown.

### PR DESCRIPTION
**Pull request by tlm**
- This change adds a new key to the Pebble service plan spec called kill-delay. This allows the author of a service definition to control the amount of time Pebble allows before sending SIGTERM and SIGKILL to a process.

- This change also increases the default duration afforded to pebble for kill-delay from 5s to 25s.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -445,6 +445,11 @@ services:
         after:
             - <other service name>
 
+        # (Optional) The amount of time afforded to this service to handle
+        # SIGTERM and exit gracefully before being force killed.
+        # Default is 25 seconds ("25s").
+        kill-delay: <grace-period>
+
```
Comment (benhoyt) [inline README.md:452]: It's weird to place this between "after" and "before". I would probably put it last, after "backoff-limit".

Also, can we clarify that this is connected to stop/restart? Perhaps:

```
        # (Optional) When a service is being stopped or restarted, this is the
        # amount of time allowed for the service to handle SIGTERM and exit
        # gracefully before being force-killed with SIGKILL. Default is 25
        # seconds ("25s").
```

File: internal/daemon/daemon.go [hunk 2]
```diff
@@ -635,10 +635,6 @@ func (d *Daemon) Stop(sigCh chan<- os.Signal) error {
 	return nil
 }
 
-// This is a little more than servstate.killWait (but that isn't and shouldn't
-// be exported, so just duplicate it here).
-const stopRunningTimeout = 5*time.Second + 100*time.Millisecond
-
```
Comment (benhoyt) [inline internal/daemon/daemon.go:641]: Do we still want to remove this? If so, why?

File: internal/overlord/servstate/handlers.go [hunk 3]
```diff
@@ -54,9 +54,15 @@ func TaskServiceRequest(task *state.Task) (*ServiceRequest, error) {
...
 
 var (
-	okayWait = 1 * time.Second
-	killWait = 5 * time.Second
-	failWait = 10 * time.Second
+	// defaultKillDelay is the duration afforded to services for processing
+	// SIGTERM signals and shutting down cleanly if the service hasn't specified
+	// their own duration.
+	defaultKillDelay = 25 * time.Second
+	okayWait         = 1 * time.Second
```
Comment (benhoyt) [inline internal/overlord/servstate/handlers.go:61]: Because `okayWait` is quite separate (and used when starting a service), can we move it to the top of this block? And perhaps add a brief comment while we're here:

```go
var (
    // okayWait is the time to wait after starting a service before we conclude
    // that it's running successfully.
    okayWait = 1 * time.Second
    ...
)
```

File: internal/overlord/servstate/manager_test.go [hunk 4]
```diff
@@ -320,6 +328,32 @@ func (s *S) stopTestServicesAlreadyDead(c *C) {
 	s.st.Unlock()
 }
 
+func (s *S) TestGracePeriodIsUsed(c *C) {
```
Comment (benhoyt) [inline internal/overlord/servstate/manager_test.go:331]: This test takes 10 seconds to run, which more than doubles our test run time. :-) Can we shorten it to a couple hundred milliseconds at most? But we also need to ensure it's not flakey.

**Review by benhoyt — COMMENTED** [review]
Looks good to me overall. Couple of nit comments and a couple of questions.

File: README.md (hunk 1 above)
Comment (benhoyt) [reply README.md:452]: Oh, one more thing: Instead of saying `<grace-period>` as the value we should say `<duration>` (like other duration fields).

File: internal/daemon/daemon.go (hunk 2 above)
Comment (tlm) [reply internal/daemon/daemon.go:641]: Because it triggers well before the service is stopped and then forces the reaper to be stopped. Doesn't make a huge amount of sense in my opinion.

Comment (tlm) [reply internal/daemon/daemon.go:641]: Try that out based on our discussions.

File: internal/overlord/servstate/manager.go [hunk 5]
```diff
@@ -300,6 +300,24 @@ func (m *ServiceManager) Services(names []string) ([]*ServiceInfo, error) {
...
+	defer m.servicesLock.Unlock()
+
+	maxDuration := defaultKillDelay
+	for _, svc := range m.services {
+		if svc.killDelay() > maxDuration {
+			maxDuration = svc.killDelay()
+		}
+	}
+
+	// We add a millisecond on to the maximum here to allow for signals to be
```
Comment (benhoyt) [inline internal/overlord/servstate/manager.go:316]: Nit: it's 100 milliseconds, so comment doesn't match code. What about just commenting "We add a little..."

**Review by benhoyt — APPROVED** [review]
Looks good to me now, thank you!

**Comment by paulomach** [comment]
Just passing by to say that's super critical for data platform deployments and we are looking forward to it. :grimacing:

**Review by delgod — APPROVED** [review]

File: internal/overlord/servstate/manager_test.go [hunk 6]
```diff
@@ -46,9 +46,9 @@ import (
 )
 
 const (
-	shortOkayWait = 50 * time.Millisecond
-	shortKillWait = 100 * time.Millisecond
-	shortFailWait = 200 * time.Millisecond
+	shortOkayWait         = 50 * time.Millisecond
+	shortKillDelay        = 100 * time.Millisecond
+	shortKillWaitDuration = 100 * time.Millisecond
```
Comment (niemeyer) [inline internal/overlord/servstate/manager_test.go:51]: Needs updating to reflect previous points.

File: internal/overlord/servstate/manager_test.go [hunk 7]
```diff
@@ -320,6 +328,51 @@ func (s *S) stopTestServicesAlreadyDead(c *C) {
 	s.st.Unlock()
 }
 
+func (s *S) TestMaximumStopRunningDuration(c *C) {
```
Comment (niemeyer) [inline internal/overlord/servstate/manager_test.go:331]: TestStopTimeout after previous points.

File: internal/overlord/servstate/manager_test.go [hunk 8]
```diff
@@ -320,6 +328,51 @@ func (s *S) stopTestServicesAlreadyDead(c *C) {
...
+	_, _, err = s.manager.Replan()
+	c.Assert(err, IsNil)
+	s.startServices(c, []string{"test9"}, 1)
+	s.waitUntilService(c, "test9", func(svc *servstate.ServiceInfo) bool {
+		return svc.Current == servstate.StatusActive
+	})
+	c.Assert(s.manager.MaximumStopRunningDuration(), Equals, time.Minute*60+time.Millisecond*100)
+}
+
+func (s *S) TestKillWaitIsUsed(c *C) {
```
Comment (niemeyer) [inline internal/overlord/servstate/manager_test.go:350]: KillDelay

File: internal/overlord/servstate/manager.go [hunk 9]
```diff
@@ -300,6 +300,24 @@ func (m *ServiceManager) Services(names []string) ([]*ServiceInfo, error) {
...
 
+// MaximumStopRunningDuration returns the worst case duration that will have to
+// be waited for to have all services in this manager stopped.
+func (m *ServiceManager) MaximumStopRunningDuration() time.Duration {
+	m.servicesLock.Lock()
+	defer m.servicesLock.Unlock()
+
+	maxDuration := defaultKillDelay
+	for _, svc := range m.services {
+		if svc.killDelay() > maxDuration {
```
Comment (niemeyer) [inline internal/overlord/servstate/manager.go:311]: A few comments here:

1. Shouldn't we be looking only at `s.state == stateRunning` here?
2. Other code that look into the map seem to consider the case where the map value is nil, which I assume means a quick shutdown. Do we have a panic waiting to happen here?
3. The code in this file (and in tests?) doesn't seem to use the `svc` naming. It's nice to keep these consistent when possible (or more consistent, if there are already multiple cases).

File: internal/overlord/servstate/manager_test.go [hunk 10]
```diff
@@ -167,7 +175,7 @@ func (s *S) SetUpTest(c *C) {
 
 	restore := servstate.FakeOkayWait(shortOkayWait)
 	s.AddCleanup(restore)
-	restore = servstate.FakeKillWait(shortKillWait, shortFailWait)
+	restore = servstate.FakeKillDelayWait(shortKillDelay, shortKillWaitDuration)
```
Comment (niemeyer) [inline internal/overlord/servstate/manager_test.go:178]: That emphasizes one of the earlier points: the delta is "kill" vs. "fail", not "wait" vs. "delay". vs. "duration

**Review by niemeyer — CHANGES_REQUESTED** [review]
Looks very nice, Thomas. Quite a few comments here, but almost all of them are cosmetic. Maybe one item worth pondering on.

File: internal/overlord/servstate/handlers.go [hunk 11]
```diff
@@ -592,6 +601,17 @@ func (s *serviceData) sendSignal(signal string) error {
...
 
+// killDelay reports the duration that this service should be given when being
+// asked to shutdown gracefully before being force terminated. The value
+// returned will either be the services pre configured value or the default
+// grace period for pebble.
+func (s *serviceData) killDelay() time.Duration {
+	if s.config.KillDelay.IsSet {
+		return s.config.KillDelay.Value
+	}
+	return defaultKillDelay
```
[the remark this answers is not in the document]
Comment (tlm) [reply internal/overlord/servstate/handlers.go:612]: Ta, I have removed the reference to grace period in the paragraph and replaced with kill delay to keep the language consistent.

File: internal/overlord/servstate/manager.go (hunk 9 above)
Comment (tlm) [reply internal/overlord/servstate/manager.go]: Thanks, I have re-worked the code a little and updated tests.

I am considering starting, running and terminating.

Terminating might seem like an odd choice my justification is if you have service a that has a kill delay of 5 minutes and you tell it to stop then immediately stop Pebble the next wait time will be what ever the maximum is of the running services we still might fall outside of that 5 minute window.

Hope that makes sense.

**Comment by beliaev-maksim** [comment]
hi @niemeyer @tlm !
looks like all the comments were addressed, are we good to merge ?

File: README.md [hunk 12]
```diff
@@ -510,6 +510,10 @@ services:
         # half a minute ("30s").
         backoff-limit: <duration>
 
+        # (Optional) The amount of time afforded to this service to handle
+        # SIGTERM and exit gracefully before SIGKILL terminates it forcefully.
+        # Default is 25 seconds ("25s").
```
Comment (niemeyer) [inline README.md:515]: This feels like too much for a default. Imagine someone sitting on the terminal for 25 seconds while they run "pebble stop foo"... they will likely CTRL-C out before the job is done as it seems stuck anyway.

File: internal/daemon/daemon.go [hunk 13]
```diff
@@ -635,10 +635,6 @@ func (d *Daemon) Stop(sigCh chan<- os.Signal) error {
 	return nil
 }
 
-// This is a little more than servstate.killWait (but that isn't and shouldn't
-// be exported, so just duplicate it here).
-const stopRunningTimeout = 5*time.Second + 100*time.Millisecond
```
Comment (niemeyer) [inline internal/daemon/daemon.go:640]: That's where the 5 seconds came from, btw.  It's a reasonable time to be waiting for an action to get done before believing that it's actually dead.

File: internal/overlord/servstate/handlers.go [hunk 14]
```diff
@@ -54,9 +54,18 @@ func TaskServiceRequest(task *state.Task) (*ServiceRequest, error) {
...
-	killWait = 5 * time.Second
-	failWait = 10 * time.Second
+	// okayDelay is the time to wait after starting a service before we conclude
+	// that it's running successfully.
+	okayDelay = 1 * time.Second
+
+	// killDelayDefault is the duration afforded to services for processing
+	// SIGTERM signals and shutting down cleanly if the service hasn't specified
+	// their own duration.
+	killDelayDefault = 25 * time.Second
```
Comment (niemeyer) [inline internal/overlord/servstate/handlers.go:64]: Per earlier note, seems high for the default.

File: internal/overlord/servstate/manager.go [hunk 15]
```diff
@@ -300,6 +300,30 @@ func (m *ServiceManager) Services(names []string) ([]*ServiceInfo, error) {
...
+		case stateStarting, stateRunning, stateTerminating:
+			if service.killDelay() > maxDuration {
+				maxDuration = service.killDelay()
+			}
+		}
+	}
+
+	// We add a little extra time here to allow for signals to be sent and
+	// processed.
+	return maxDuration + 100*time.Millisecond
```
Comment (niemeyer) [inline internal/overlord/servstate/manager.go:324]: We need to add failDelay here, as that's the amount of time we're willing to wait before the service is supposed to actually be terminated. Otherwise, 100 milliseconds is way too short and may generate bugs due to the process dying while it's still attempting to deliver the final KILL signals.

File: internal/overlord/servstate/manager_test.go [hunk 16]
```diff
@@ -46,9 +46,9 @@ import (
 )
 
 const (
-	shortOkayWait = 50 * time.Millisecond
-	shortKillWait = 100 * time.Millisecond
-	shortFailWait = 200 * time.Millisecond
+	shortOkayDelay = 50 * time.Millisecond
+	shortKillDelay = 100 * time.Millisecond
+	shortFailDelay = 100 * time.Millisecond
```
Comment (niemeyer) [inline internal/overlord/servstate/manager_test.go:51]: Why was this changed?

**Review by niemeyer — CHANGES_REQUESTED** [review]
All the logic looks fine, thank you.

Only points remaining are about the tuning of times.

File: internal/overlord/servstate/manager_test.go (hunk 16 above)
Comment (benhoyt) [reply internal/overlord/servstate/manager_test.go:51]: It's because the original `shortFailWait` included both the time from first SIGTERM to SIGKILL and the second wait after that (100+100=200). Now the code is such that "fail delay" only includes that second part (which I think is a bit clearer in any case). So the total wait time here wasn't changed.

File: README.md (hunk 12 above)
Comment (tlm) [reply README.md:515]: This was changed to 25s to be inline with Kubernetes default 30s. Allowing enough time by default for processes that require a little extra time. My understanding was this was an agreed upon item.

However happy to change. Should we take this back to 5s or another value? @benhoyt do you have any thoughts on what this value should be?

File: internal/overlord/servstate/manager.go (hunk 15 above)
Comment (tlm) [reply internal/overlord/servstate/manager.go:324]: Thanks good pick up!

File: internal/daemon/daemon.go (hunk 13 above)
Comment (tlm) [reply internal/daemon/daemon.go:640]: Thanks have adjusted below.

File: internal/overlord/servstate/handlers.go (hunk 14 above)
Comment (tlm) [reply internal/overlord/servstate/handlers.go:64]: As per above.

File: README.md (hunk 12 above)
Comment (jnsgruk) [reply README.md:515]: I think the point is that in cases where there is more direct user interaction, 25s feels too long, where 5s doesn't.

Presumably this can still be set to 25/30s when Juju writes the plan for the unit agent either way, and still solve our original issue?

Comment (benhoyt) [reply README.md:515]: Yeah, agreed re a user interacting with `pebble stop`. I reckon let's keep the default 5s and Juju can easily override.

Comment (tlm) [reply README.md:515]: Easy done.

File: internal/overlord/servstate/manager_test.go (hunk 16 above)
Comment (niemeyer) [reply internal/overlord/servstate/manager_test.go:51]: Ack, thanks.

**Review by niemeyer — APPROVED** [review]
Thanks for the collaboration here, everyone.
